src/socket/udp_socket.cr

```diff
@@ -88,4 +88,142 @@ class UDPSocket < IPSocket
+  # destination group and multicast loopback is enabled.
+  def multicast_hops
+    case @family
+    when Family::INET
+      getsockopt LibC::IP_MULTICAST_TTL, 0, LibC::IPPROTO_IP
+    when Family::INET6
+      getsockopt LibC::IPV6_MULTICAST_HOPS, 0, LibC::IPPROTO_IPV6
+    else
+      raise Socket::Error.new "Unsupported IP address family: #{@family}"
+    end
+  end
+
+  # The multicast hops option controls the `hoplimit` field on uni-cast packets.
+  # If `-1` is specified, the kernel will use a default value.
+  # If a value of `0` to `255` is specified, the packet will have the specified
+  # value as `hoplimit`. Other values are considered invalid and `Errno` will be raised.
+  # Datagrams with a `hoplimit` of `1` are not forwarded beyond the local network.
+  # Multicast datagrams with a `hoplimit` of `0` will not be transmitted on any
+  # network, but may be delivered locally if the sending host belongs to the
+  # destination group and multicast loopback is enabled.
+  def multicast_hops=(val : Int)
+    case @family
+    when Family::INET
+      setsockopt LibC::IP_MULTICAST_TTL, val, LibC::IPPROTO_IP
+    when Family::INET6
+      setsockopt LibC::IPV6_MULTICAST_HOPS, val, LibC::IPPROTO_IPV6
+    else
+      raise Socket::Error.new "Unsupported IP address family: #{@family}"
+    end
+    val
+  end
+
+  # For hosts with multiple interfaces, each multicast transmission is sent
+  # from the primary network interface. This function overrides the default
+  # IPv4 interface address for subsequent transmissions. Setting the interface
+  # to `0.0.0.0` will select the default interface.
+  # Raises `Socket::Error` unless the socket is IPv4 and an IPv4 address is provided.
+  def multicast_interface(address : IPAddress)
+    if @family == Family::INET
+      if addr = address.addr4
+        setsockopt LibC::IP_MULTICAST_IF, addr, LibC::IPPROTO_IP
+      else
+        raise Socket::Error.new "Expecting an IPv4 interface address. Address provided: #{address.address}"
+      end
+    else
+      raise Socket::Error.new "For use with IPv4 only. Unsupported IP address family: #{@family}"
+    end
+  end
+
+  # For hosts with multiple interfaces, each multicast transmission is sent
+  # from the primary network interface. This function overrides the default
+  # IPv6 interface for subsequent transmissions. Setting the interface to
+  # index `0` will select the default interface.
+  def multicast_interface(index : UInt32)
+    if @family == Family::INET6
+      setsockopt LibC::IPV6_MULTICAST_IF, index, LibC::IPPROTO_IPV6
+    else
+      raise Socket::Error.new "For use with IPv6 only. Unsupported IP address family: #{@family}"
+    end
+  end
+
+  # A host must become a member of a multicast group before it can receive
+  # datagrams sent to the group.
+  # Raises `Socket::Error` if an incompatible address is provided.
+  def join_group(address : IPAddress)
+    group_modify(address, true)
```

Comment:
It would be ore concise if `operation` directly takes the appropriate value (e.g. `LibC::IPV6_JOIN_GROUP`) instead of an indirection over a boolean flag.